signcrypt_open.go

```diff
@@ -0,0 +1,337 @@
+	}
+	return n, err
+}
+
+func (sos *signcryptOpenStream) read(b []byte) (n int, err error) {
+
+	// Handle the case of a previous error. Just return the error
+	// again.
+	if sos.err != nil {
+		return 0, sos.err
+	}
+
+	// Handle the case first of a previous read that couldn't put all
+	// of its data into the outgoing buffer.
+	if len(sos.buf) > 0 {
+		n = copy(b, sos.buf)
+		sos.buf = sos.buf[n:]
+		return n, nil
+	}
+
+	// We have two states we can be in, but we can definitely
+	// fall through during one read, so be careful.
+
+	if sos.state == stateBody {
+		var last bool
+		n, last, sos.err = sos.readBlock(b)
+		if sos.err != nil {
+			return 0, sos.err
+		}
+
+		if last {
+			sos.state = stateEndOfStream
+		}
+	}
+
+	if sos.state == stateEndOfStream {
+		sos.err = assertEndOfStream(sos.mps)
+		if sos.err != nil {
+			return 0, sos.err
+		}
+	}
+
+	return n, nil
+}
+
+func (sos *signcryptOpenStream) readHeader(rawReader io.Reader) error {
+	// Read the header bytes.
+	headerBytes := []byte{}
+	seqno, err := sos.mps.Read(&headerBytes)
+	if err != nil {
+		return ErrFailedToReadHeaderBytes
+	}
+	// Compute the header hash.
+	headerHash := sha512.Sum512(headerBytes)
+	sos.headerHash = headerHash[:]
+	// Parse the header bytes.
+	var header SigncryptionHeader
+	err = decodeFromBytes(&header, headerBytes)
+	if err != nil {
+		return err
+	}
+	header.seqno = seqno
+	err = sos.processSigncryptionHeader(&header)
+	if err != nil {
+		return err
+	}
+	sos.state = stateBody
+	return nil
+}
+
+func (sos *signcryptOpenStream) readBlock(b []byte) (n int, lastBlock bool, err error) {
+	var sb signcryptionBlock
+	var seqno packetSeqno
+	seqno, err = sos.mps.Read(&sb)
+	if err != nil {
+		return 0, false, err
+	}
+	sb.seqno = seqno
+	var plaintext []byte
+	plaintext, err = sos.processSigncryptionBlock(&sb)
+	if err != nil {
+		return 0, false, err
+	}
+	if plaintext == nil {
+		return 0, true, err
+	}
+
+	// Copy as much as we can into the given outbuffer
+	n = copy(b, plaintext)
+	// Leave the remainder for a subsequent read
+	sos.buf = plaintext[n:]
+
+	return n, false, err
+}
+
+func (sos *signcryptOpenStream) tryBoxSecretKeys(hdr *SigncryptionHeader, ephemeralPub BoxPublicKey) (*SymmetricKey, error) {
+	derivedKeys := []*SymmetricKey{}
+	for _, receiverBoxSecretKey := range sos.keyring.GetAllBoxSecretKeys() {
+		sharedSecretBox := receiverBoxSecretKey.Box(ephemeralPub, nonceForDerivedSharedKey(), make([]byte, 32))
+		derivedKey, err := symmetricKeyFromSlice(sharedSecretBox[len(sharedSecretBox)-32 : len(sharedSecretBox)])
+		if err != nil {
+			panic(err) // should be statically impossible, if the slice above is the right length
+		}
+		derivedKeys = append(derivedKeys, derivedKey)
+	}
+
+	// Try each of the box secret keys against each of the receiver pairs in
+	// the message header. The actual expected number of box secret keys is
+	// one, so this shouldn't be as quadratic as it looks.
+	for receiverIndex, receiver := range hdr.Receivers {
```

Comment:
It could be in a different application, but in our case "all keys" is your one device key. (The large number of TLF symmetric keys out there is why we have the server-assisted pseudonym mechanism, to avoid blowing up this loop.) It might increase if we introduce shared master keys, but I think in general it'll stay reasonable.

The reason we have to recompute the identifier for each position is that we want to guarantee that two users with the same public key won't leak that fact. We _could_ instead put the responsibility on the implementation to deduplicate them. The spec might be able to help with that by making it a map of recipients instead of a string...though Go really hates maps with []byte keys...